File: src/bot/commands/player/upgrades.ts
import { SlashCommandBuilder, type ChatInputCommandInteraction } from 'discord.js';
import { Command } from '../../struct/Command.js';
import { baseEmbed } from '../../../util/embeds.js';
import { pad, padStart } from '../../../util/helper.js';

/** Lists units that are not yet at their current max level (i.e. available to upgrade). */
export default class UpgradesCommand extends Command {
  public constructor() {
    super('upgrades', { category: 'player', description: 'List a player’s available upgrades.', defer: true });
  }

  public builder() {
    return new SlashCommandBuilder()
      .setName('upgrades')
      .setDescription('List a player’s available upgrades.')
      .addStringOption((opt) => opt.setName('tag').setDescription('Player tag. Defaults to your linked account.'))
      .addUserOption((opt) => opt.setName('user').setDescription('Show another linked user.'));
  }

  private table(items: { name: string; level: number; maxLevel: number }[]): string {
    const pending = items.filter((u) => u.level < u.maxLevel);
    if (!pending.length) return '_all maxed_';
    return pending
      .map((u) => `\`${pad(u.name, 18)} ${padStart(u.level, 2)} → ${padStart(u.maxLevel, 2)}\``)
      .join('\n')
      .slice(0, 1024);
  }

  public async exec(interaction: ChatInputCommandInteraction, args: Record<string, any>) {
    const { data: player, error } = await this.client.resolver.resolvePlayer(interaction, args.tag, args.user);
    if (!player) return interaction.editReply(error!);

    const home = <T extends { village: string }>(arr: T[]) => arr.filter((x) => x.village === 'home');
    const embed = baseEmbed(this.client, interaction)
      .setTitle(`${player.name} (${player.tag}) — Upgrades Available`)
      .addFields(
        { name: 'Heroes', value: this.table(home(player.heroes)) },
        { name: 'Troops', value: this.table(home(player.troops)) },
        { name: 'Spells', value: this.table(home(player.spells)) }
      );

    return interaction.editReply({ embeds: [embed] });
  }
}

File: cogs/player/commands/upgrades.py
"""/player upgrades, units not yet at their current max level."""

import discord

from utils.embeds import base_embed, error_embed
from utils.emojis import get_unit_emoji
from utils.helpers import pad, pad_start
from utils.resolver import resolve_player
from utils.units import home_units


def _pending(items) -> str:
    pending = [u for u in items if u["level"] < u["maxLevel"]]
    if not pending:
        return "_all maxed_"
    lines = []
    for u in pending:
        e = get_unit_emoji(u["name"])
        prefix = f"{e} " if e else ""
        lines.append(f"{prefix}`{pad(u['name'], 16)} {pad_start(u['level'], 2)} > {pad_start(u['maxLevel'], 2)}`")
    return "\n".join(lines)[:1024]


async def upgrades(interaction: discord.Interaction, tag: str = None, user: discord.User = None):
    await interaction.response.defer()
    try:
        player = await resolve_player(interaction, tag, user)
    except ValueError as err:
        await interaction.followup.send(embed=error_embed(str(err)))
        return

    embed = await base_embed(interaction, title=f"{player.name} ({player.tag}), Upgrades Available")
    embed.add_field(name="Heroes", value=_pending(home_units(player, "heroes")), inline=False)
    embed.add_field(name="Troops", value=_pending(home_units(player, "troops")), inline=False)
    embed.add_field(name="Spells", value=_pending(home_units(player, "spells")), inline=False)

    await interaction.followup.send(embed=embed)
